.ci/benchmarks.groovy

```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/env groovy
+@Library('apm@current') _
+
+pipeline {
+  agent { label 'linux && immutable && obs11' }
+  environment {
+    REPO = 'apm-server'
+    BASE_DIR = "src/github.com/elastic/${env.REPO}"
+    BRANCH_NAME_LOWER_CASE = "${env.BRANCH_NAME.toLowerCase().replaceAll('[^a-z0-9-]', '-')}"
+    AWS_ACCOUNT_SECRET = 'secret/observability-team/ci/elastic-observability-aws-account-auth'
+    GCP_ACCOUNT_SECRET = 'secret/observability-team/ci/elastic-observability-account-auth'
+    EC_KEY_SECRET = 'secret/observability-team/ci/elastic-cloud/observability-pro'
+    TERRAFORM_VERSION = '1.1.9'
+
+    CREATED_DATE = "${new Date().getTime()}"
+    JOB_GCS_BUCKET_STASH = 'apm-ci-temp'
+    JOB_GCS_CREDENTIALS = 'apm-ci-gcs-plugin'
+  }
+
+  options {
+    timeout(time: 2, unit: 'HOURS')
+    buildDiscarder(logRotator(numToKeepStr: '100', artifactNumToKeepStr: '30', daysToKeepStr: '30'))
+    timestamps()
+    ansiColor('xterm')
+    disableResume()
+    durabilityHint('PERFORMANCE_OPTIMIZED')
+    rateLimitBuilds(throttle: [count: 60, durationName: 'hour', userBoost: true])
+    quietPeriod(10)
+  }
+
+  stages {
+    stage('Checkout') {
+      options { skipDefaultCheckout() }
+      steps {
+        deleteDir()
+        gitCheckout(basedir: "${BASE_DIR}", shallow: false)
+      }
+    }
+
+    stage('Benchmarks') {
+      options { skipDefaultCheckout() }
+      steps {
+        dir ("${BASE_DIR}") {
+          withGoEnv() {
+            dir("testing/benchmark") {
+              withTestClusterEnv {
+                  sh(label: 'Build apmbench', script: 'make apmbench $SSH_KEY terraform.tfvars')
+                  sh(label: 'Spin up benchmark environment', script: 'make init apply')
+                  sh(label: 'Run benchmarks', script: 'make run-benchmark index-benchmark-results')
+                }
+              }
+            }
+          }
+        }
+      }
+      post {
+        always {
+          dir("${BASE_DIR}/testing/benchmark") {
+            stashV2(name: 'build', bucket: "${JOB_GCS_BUCKET_STASH}", credentialsId: "${JOB_GCS_CREDENTIALS}")
+            withTestClusterEnv {
+              sh(label: 'Tear down benchmark environment', script: 'make destroy')
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+def withTestClusterEnv(Closure body) {
+  withGCPEnv(secret: "${GCP_ACCOUNT_SECRET}") {
+    withAWSEnv(secret: "${AWS_ACCOUNT_SECRET}") {
+      withTerraformEnv(version: "${TERRAFORM_VERSION}") {
+        withECKey {
+          withEnv(['SSH_KEY=./id_rsa_terraform',
```

Comment:
@marclop Is the preferable way you'd like to tweak terraform variables?

Reply:
I'm afraid I don't know which other ways exist to tweak variables. I doubt we'd want to change the value for `SSH_KEY`, but we will want to change the value of a few of the existing Makefile variables, for instance the `GOBENCH_*` variables should be most likely configured statically on the job (secrets + values derived from branch) https://github.com/elastic/apm-server/blob/3cf9bdb24ea5f3158a10cded5888151697f517f2/testing/benchmark/Makefile#L12-L15

* `GOBENCH_USERNAME`: Shouldn't necessarily be change per job, but be configured with the relevant secret to be able to ingest the benchmark results to the `observability-benchmarks` cluster.
* `GOBENCH_PASSWORD`: Same.
* `GOBENCH_INDEX`: Most likely we'll need to use an index that resembles: `gobench-v2-<BranchName>`. This is to allow us ingesting the daily / nightly benchmarks in `gobench-v2-main`, and going forward, trigger on-demmand benchmarks on PRs and ingest the results in a different index: `gobench-v2-pr_xyz`
* `GOBENCH_HOST`: The Elasticsearch URL of observability-benchmarks.
* `GOBENCH_TAGS`: Should contain a series of `k=v` that helps us identify the configuration of the benchmark that was run such as; APM Server size, Elasticsearch size, etc. (See below)

https://github.com/elastic/apm-server/blob/3cf9bdb24ea5f3158a10cded5888151697f517f2/testing/benchmark/Makefile#L73-L75

Last, the actual parameters for `apmbench` so we can configure it differently for different sizes of benchmarks. Are these things configurable as inputs in Jenkins? We can have defaults for the job, and perhaps have different jobs for different sizes.

https://github.com/elastic/apm-server/blob/3cf9bdb24ea5f3158a10cded5888151697f517f2/testing/benchmark/Makefile#L5-L10